### PR TITLE
[WIP] base contentproviders commit

### PR DESCRIPTION
**Pull request by yuvipanda**
Very much a WIP, but some thougts on how we can support more
than just git.

### CODE REVIEW

**Comment by yuvipanda** [comment]
This can be used for both #96 and #102

**Comment by yuvipanda** [comment]
It would be great if someone can take this on and push it to completion :)

File: repo2docker/contentproviders/git.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+class GitContentProvider(ContentProvider):
+    """Provides contents of a git repository (optionally at a given ref)
+    """
+    kind = "git"
+
+    def provide(self, spec, output_dir, yield_output=False):
+        url = spec['url']
+        ref = spec.get('ref', None)
+        try:
+            for line in execute_cmd(['git', 'clone', url, output_dir],
```
Comment (minrk) [inline repo2docker/contentproviders/git.py]: From #233 add '--recursive' here

File: repo2docker/contentproviders/git.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            raise ContentProviderException("Failed to clone repository!") from e
+
+        if ref:
+            try:
+                for line in execute_cmd(['git', 'reset', '--hard', ref],
+                                        cwd=output_dir,
+                                        capture=yield_output):
+                    yield line
+            except subprocess.CalledProcessError:
+                raise ContentProviderException("Failed to checkout ref {}!".format(ref)) from e
```
Comment (minrk) [inline repo2docker/contentproviders/git.py]: Missing reference for `from e` here

**Comment by yuvipanda** [comment]
Superseeded by #242

**Pull request closed without merging**